SCLang: avoid collecting inside primitives

JordanHendersonMusic opened the pull request:
<!-- For information about contributing see: https://github.com/supercollider/supercollider/wiki/Contributing-directory -->

## Purpose and Motivation

When you allocate inside a primitive, if you don't connect that resource to the memory graph the next time you allocate (and call collect) the object might be collected, leaving you with an invalid memory resource. This has caused many bugs in the past. This PR simply holds off calling collect until the primitive has finished.

This has the potential draw back that if you are creating many temporary objects, you might run out of memory. But none of the primitives do this right now, and the ones that do allocate a lot of temporary objects usually disable the collection cycle anyway to avoid having the GC free the objects before they are placed on the stack.

## Types of changes

<!-- Delete lines that don't apply -->

- Memory management improvement


## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review <!-- If not ready for review, consider making a Draft PR instead, or describe what kind of specific feedback you want. -->

<!-- ## Merge notes
(optional) You may also add notes of:
     - Dependencies: if this PR depends on any other PRs or actions
     - Merge Note: any considerations regarding merging this PR
     - Release Notes: text you think would be useful to include in the Release Notes
-->

<!-- Note: The recommended commit message format is:
     component: sub-component: short description of changes
For example:
     classlib: ClassBrowser: fix search with empty query string
Common `comp:` labels:
     help, classlib, sclang, scsynth, plugins, tests, docs, SCDoc
See wiki for more info: Wiki->Creating Pull Requests->git-commit-message-format
-->

JordanHendersonMusic committed:
SCLang: avoid collecting inside primitives
Wait until the primitive has finished to call collect.
This has the potential draw back that if you are creating many temporary objects, you might run out of memory. But none of the primitives do this right now, and the ones that do allocate a lot of temporary objects usually disable the collection cycle anyway to avoid having the GC free the objects before they are placed on the stack.

Spacechild1 commented:
I like this!

> When you allocate inside a primitive, if you don't connect that resource to the memory graph the next time you allocate (and call collect) the object might be collected, leaving you with an invalid memory resource. 

This is indeed a dangerous pitfall! Not disabling the GC for temporary objects is an easy mistake to make. It's just a Boolean argument in `newPyrObject`, `newPyrString`, `newPyrArray`, etc. and as such it's really easy to overlook. I've made this mistake myself in the past and only discovered it by accident.

JordanHendersonMusic commented:
Yes it is easy to do! 

I've been writing a lot of zig code recently which forces you to think in terms of allocators and to find patterns of memory lifetimes, which made me realise if primitives used something like an area allocator this issue goes away. 

There was a proposal (@muellmusik I think you mentioned this once?) to model this relationship using c++'s object model, like you would in rust, but I think it is a bit overkill. 

You do technically lose the ability to free temporary objects made inside the primitive with this approach,  but SC usually has plenty of free memory available as the startup process allocates a lot and frees it once done, and we don't tend to allocate too much inside the existing primitives anyway.

Spacechild1 commented:
> You do technically lose the ability to free temporary objects made inside the primitive with this approach,

We can always provide an escape hatch for people who (think they) know what there are doing. E.g. a function like `enableGCinPrimitive`.

JordanHendersonMusic commented:
Sure, nice and easy, it can just disable the 'insidePrimitive' bool and be called at the top of the primitive.

JordanHendersonMusic committed:
Rename and add an escape hatch

JordanHendersonMusic commented:
Okay I've added that, I've also renamed these methods because it doesn't seem right that the GC knows about primitives, it is now referred to as a 'delayed collection context'.

JordanHendersonMusic committed:
typo

muellmusik commented:
Hey @JordanHendersonMusic! At a quick glance this looks great! My favourite type of solution: One that just handles it for you. 

My only caveat (besides test test test) would be that it's worth considering this within a context of a broader rework of allocation, which I think would still be a good idea as there are a number of similar pitfalls. That said, I can't see any particular issues with that based on a quick look, but it's perhaps worth thinking through.

> @muellmusik I think you mentioned this once?

Yup, I think this was the last version of a PR attempting to look at this broadly: https://github.com/supercollider/supercollider/pull/3384

I laid out my understanding of the problems to address here: https://github.com/supercollider/supercollider/issues/1634#issuecomment-354433378

JordanHendersonMusic commented:
Thanks!

Re testing...

I've checked this does indeed call collect when leaving a primitive, and have checked there aren't any memory leaks directly resulting from this code. I don't really know how else to test it as the worst that can happen is that collect isn't called until the next primitive call. I think of this change as meaning we no longer have to check all the primitives for this use-after-unintended-free bug, but instead just have to check the delayed  collect works (which it does).

Re the other issues raised in that comment, I think a lot of them aren't really related to allocation directly, and can instead be fixed through helper methods. I've got a bunch of ideas about this... But I don't like any of them because they all involve complex c++ concepts (or worse, macros) which I'm really keen to avoid.

JordanHendersonMusic commented:
I think this is good to be reviewed, and potentially merged if no one sees any issues. I've done some more testing and can't find any issues, so it really just needs to be tested in the development branch.

muellmusik commented:
Would it be worth it to add something to the test suite?

Otherwise I think this looks really good.

JordanHendersonMusic commented:
The only way I can think to test this, which is what I have been doing, is to run a bunch of code (I've been running parts of the sclang unit test suite) and check the memory usage doesn't grow. I don't think that is something that can be reasonably tested using our existing tooling, but I'm happy for suggestions!

muellmusik commented:
> I don't think that is something that can be reasonably tested using our existing tooling, but I'm happy for suggestions!

I guess what would be nice would be some test that using the existing or new gc introspection methods just checks that it's working as predicted in a range of cases (i.e. in and out of collections, etc). I suspect that could be a little finicky to get right though, and I don't want it to hold up this PR.

JordanHendersonMusic commented:
I don't think I'll add anything like that in this pr. But in future I think the GC part of the code base could be reworked to allow for proper unit tests.

JordanHendersonMusic commented:
@Spacechild1 do you think this is good too? If so I'll merge.

Spacechild1 commented on lang/LangSource/GC.h:
```diff
@@ -207,6 +229,9 @@ class PyrGC {
     bool mRunning;
+
+    bool delayCollection { false };
+    bool attemptedToCollectWhenDelayed { false };
```
I think these should have a `m` prefix like all the other member variables.

Spacechild1 commented:
Just a small stylistic thing. Otherwise this looks good!

JordanHendersonMusic committed:
rename members
